File: app/src/main/java/com/aexon/theme/hct/AexonHctCompat.java
package com.aexon.theme.hct;

import androidx.annotation.ColorInt;

import com.aexon.theme.hct.utils.AexonColorUtils;

public class AexonHctCompat {
	private final double hue;
	private final double chroma;
	private final double tone;
	
	public AexonHctCompat(double hue, double chroma, double tone) {
		this.hue = hue;
		this.chroma = chroma;
		this.tone = tone;
	}
	
	public double getHue() {
		return hue;
	}
	
	public double getChroma() {
		return chroma;
	}
	
	public double getTone() {
		return tone;
	}
	
	@ColorInt
	public int toInt() {
		return toInt(tone);
	}
	
	@ColorInt
	public int toInt(double targetTone) {
		if (targetTone <= 0.0) return 0xFF000000;
		if (targetTone >= 100.0) return 0xFFFFFFFF;
		
		double maxSafeChroma = findMaxSafeChromaAtTone(hue, targetTone);
		double actualChroma = Math.min(chroma, maxSafeChroma);
		
		double a = actualChroma * Math.cos(Math.toRadians(hue));
		double b = actualChroma * Math.sin(Math.toRadians(hue));
		return AexonColorUtils.argbFromLab(targetTone, a, b);
	}
	
	public static AexonHctCompat fromInt(@ColorInt int argb) {
		double[] lab = AexonColorUtils.labFromArgb(argb);
		double l = lab[0];
		double a = lab[1];
		double b = lab[2];
		
		double hue;
		double chroma;
		
		if (l < 8.0) {
			int r = (argb >> 16) & 0xFF;
			int g = (argb >> 8) & 0xFF;
			int bl = argb & 0xFF;
			int max = Math.max(r, Math.max(g, bl));
			if (max > 0) {
				double nr = r / (double) max;
				double ng = g / (double) max;
				double nb = bl / (double) max;
				int boosted = AexonColorUtils.argbFromRgb((int) Math.round(nr * 180), (int) Math.round(ng * 180), (int) Math.round(nb * 180));
				double[] boostedLab = AexonColorUtils.labFromArgb(boosted);
				hue = Math.toDegrees(Math.atan2(boostedLab[2], boostedLab[1]));
				if (hue < 0) hue += 360.0;
				chroma = Math.sqrt(boostedLab[1] * boostedLab[1] + boostedLab[2] * boostedLab[2]);
			} else {
				hue = 0.0;
				chroma = 0.0;
			}
		} else {
			hue = Math.toDegrees(Math.atan2(b, a));
			if (hue < 0) hue += 360.0;
			chroma = Math.sqrt(a * a + b * b);
		}
		
		return new AexonHctCompat(hue, chroma, l);
	}
	
	public static double findMaxSafeChromaAtTone(double hue, double tone) {
		if (tone <= 0.0 || tone >= 100.0) return 0.0;
		double hueRad = Math.toRadians(hue);
		double cosH = Math.cos(hueRad);
		double sinH = Math.sin(hueRad);
		double low = 0.0;
		double high = 160.0;
		
		for (int i = 0; i < 24; i++) {
			double mid = (low + high) / 2.0;
			double a = mid * cosH;
			double b = mid * sinH;
			if (isLabInGamut(tone, a, b)) {
				low = mid;
			} else {
				high = mid;
			}
		}
		return low;
	}
	
	private static boolean isLabInGamut(double l, double a, double b) {
		double fy = (l + 16.0) / 116.0;
		double fx = a / 500.0 + fy;
		double fz = fy - b / 200.0;
		
		double x = labInvf(fx) * 95.047;
		double y = labInvf(fy) * 100.0;
		double z = labInvf(fz) * 108.883;
		
		double linearR = 3.2413774792388685 * x - 1.5376652402851851 * y - 0.49885366846268053 * z;
		double linearG = -0.9691452513005321 * x + 1.8758853451067872 * y + 0.04156585616912061 * z;
		double linearB = 0.05562093689691305 * x - 0.20395524564742123 * y + 1.0571799111220335 * z;
		
		double epsilon = 0.001;
		return linearR >= -epsilon && linearR <= 100.0 + epsilon && linearG >= -epsilon && linearG <= 100.0 + epsilon && linearB >= -epsilon && linearB <= 100.0 + epsilon;
	}
	
	private static double labInvf(double ft) {
		double e = 216.0 / 24389.0;
		double kappa = 24389.0 / 27.0;
		double ft3 = ft * ft * ft;
		return ft3 > e ? ft3 : (116.0 * ft - 16.0) / kappa;
	}
}

File: app/src/main/java/com/aexon/theme/hct/AexonHctCompat.kt
package com.aexon.theme.hct

import androidx.annotation.ColorInt
import com.aexon.theme.hct.utils.AexonColorUtils
import kotlin.math.atan2
import kotlin.math.cos
import kotlin.math.max
import kotlin.math.min
import kotlin.math.roundToInt
import kotlin.math.sin
import kotlin.math.sqrt

/**
* HCT (Hue-Chroma-Tone) versi Aexon — inti algoritma Material You
* buat generate skema warna dari 1 seed color.
*/
class AexonHctCompat(val hue: Double, val chroma: Double, val tone: Double) {
	
	@ColorInt
	fun toInt(): Int = toInt(tone)
	
	@ColorInt
	fun toInt(targetTone: Double): Int {
		if (targetTone <= 0.0) return 0xFF000000.toInt()
		if (targetTone >= 100.0) return 0xFFFFFFFF.toInt()
		
		// Chroma diclamp ke batas gamut sRGB paling aman di tone target,
		// biar hasil convert nggak "clipping" jadi warna yang salah
		val maxSafeChroma = findMaxSafeChromaAtTone(hue, targetTone)
		val actualChroma = minOf(chroma, maxSafeChroma)
		
		val a = actualChroma * cos(Math.toRadians(hue))
		val b = actualChroma * sin(Math.toRadians(hue))
		return AexonColorUtils.argbFromLab(targetTone, a, b)
	}
	
	companion object {
		fun fromInt(@ColorInt argb: Int): AexonHctCompat {
			val lab = AexonColorUtils.labFromArgb(argb)
			val l = lab[0]
			val a = lab[1]
			val b = lab[2]
			
			var hue: Double
			var chroma: Double
			
			// Warna sangat gelap (L* < 8) di-"boost" dulu biar hue/chroma-nya
			// nggak noise akibat presisi Lab yang jelek di area gelap
			if (l < 8.0) {
				val r = (argb shr 16) and 0xFF
				val g = (argb shr 8) and 0xFF
				val bl = argb and 0xFF
				val max = maxOf(r, maxOf(g, bl))
				if (max > 0) {
					val nr = r / max.toDouble()
					val ng = g / max.toDouble()
					val nb = bl / max.toDouble()
					val boosted = AexonColorUtils.argbFromRgb((nr * 180).roundToInt(), (ng * 180).roundToInt(), (nb * 180).roundToInt())
					val boostedLab = AexonColorUtils.labFromArgb(boosted)
					hue = atan2(boostedLab[2], boostedLab[1]).let { 
						(Math.toDegrees(it) + 360.0) % 360.0 
					}
					chroma = sqrt(boostedLab[1] * boostedLab[1] + boostedLab[2] * boostedLab[2])
				} else {
					hue = 0.0
					chroma = 0.0
				}
			} else {
				hue = atan2(b, a).let { 
					(Math.toDegrees(it) + 360.0) % 360.0 
				}
				chroma = sqrt(a * a + b * b)
			}
			
			return AexonHctCompat(hue, chroma, l)
		}
		
		// Binary search: cari chroma maksimum di hue+tone tertentu yang masih
		// bisa direpresentasikan sebagai warna sRGB valid (nggak keluar gamut)
		fun findMaxSafeChromaAtTone(hue: Double, tone: Double): Double {
			if (tone <= 0.0 || tone >= 100.0) return 0.0
			val hueRad = Math.toRadians(hue)
			val cosH = cos(hueRad)
			val sinH = sin(hueRad)
			var low = 0.0
			var high = 160.0
			
			repeat(24) {
				val mid = (low + high) / 2.0
				val a = mid * cosH
				val b = mid * sinH
				if (isLabInGamut(tone, a, b)) {
					low = mid
				} else {
					high = mid
				}
			}
			return low
		}
		
		private fun isLabInGamut(l: Double, a: Double, b: Double): Boolean {
			val fy = (l + 16.0) / 116.0
			val fx = a / 500.0 + fy
			val fz = fy - b / 200.0
			
			val x = labInvf(fx) * 95.047
			val y = labInvf(fy) * 100.0
			val z = labInvf(fz) * 108.883
			
			val linearR = 3.2413774792388685 * x - 1.5376652402851851 * y - 0.49885366846268053 * z
			val linearG = -0.9691452513005321 * x + 1.8758853451067872 * y + 0.04156585616912061 * z
			val linearB = 0.05562093689691305 * x - 0.20395524564742123 * y + 1.0571799111220335 * z
			
			val epsilon = 0.001
			return linearR >= -epsilon && linearR <= 100.0 + epsilon &&
			linearG >= -epsilon && linearG <= 100.0 + epsilon &&
			linearB >= -epsilon && linearB <= 100.0 + epsilon
		}
		
		private fun labInvf(ft: Double): Double {
			val e = 216.0 / 24389.0
			val kappa = 24389.0 / 27.0
			val ft3 = ft * ft * ft
			return if (ft3 > e) ft3 else (116.0 * ft - 16.0) / kappa
		}
	}
}